### PR TITLE
hook wiggled wasi-common up to lucet

### PR DESCRIPTION
**Pull request by pchickey**
Update lucet-wasi to use the wiggle family of tools, and the latest wasi-common (0.15).

This PR changes lucet-wasi to support WASI snapshot 1, and drops support for snapshot 0. Support for snapshot 0 will be restored in the future, when we add polyfill support to wiggle and wasi-common.

Included with this change is upgrading the rust toolchain to 1.42.0 (which targets snapshot 1 in the wasm32-wasi std) and wasi-sdk to release 10 (which targets snapshot 1).

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -28,16 +30,16 @@ RUN update-alternatives --install /usr/bin/clang clang /usr/bin/clang-6.0 100
 ENV LD_LIBRARY_PATH=/usr/local/lib
 
 # Install our supported version of Rust, rustfmt, and the wasm32-wasi cross-compilation target
-RUN curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain 1.40.0 -y
+RUN curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain 1.42.0 -y
```
Comment (fst-crenshaw) [inline Dockerfile:33]: 💯

File: lucet-wiggle/generate/src/lib.rs [hunk 2]
```diff
@@ -54,15 +58,13 @@ pub fn generate(doc: &witx::Document, config: &Config) -> TokenStream {
...
-            let ctx_type = config.wiggle.ctx.name.clone();
-            let ctx_constructor = config.constructor.clone();
             quote! {
                 #[lucet_hostcall]
                 #[no_mangle]
                 pub fn #name(vmctx: &mut lucet_runtime::vmctx::Vmctx, #(#func_args),*) -> #rets {
                     let memory= lucet_wiggle_runtime::LucetMemory::new(vmctx);
                     let mut ctx: #ctx_type = #ctx_constructor;
-                    #mod_name::#method_name(&ctx, &memory, #(#call_args),*)
+                    super::#mod_name::#method_name(&ctx, &memory, #(#call_args),*)
```
[the remark this answers is not in the document]
Comment (pchickey) [reply lucet-wiggle/generate/src/lib.rs:67]: nope, we only want to pull the types from the common wiggle_mod_path, the abi-level functions need to be the ones generated specifically for `#ctx_type`.

File: lucet-wiggle/generate/src/lib.rs [hunk 3]
```diff
@@ -29,8 +28,13 @@ pub fn bindings(doc: &witx::Document) -> Bindings {
     Bindings::new(bs)
 }
 
-pub fn generate(doc: &witx::Document, config: &Config) -> TokenStream {
-    let names = wiggle_generate::Names::new(&config.wiggle);
+pub fn generate(
```
[the remark this answers is not in the document]
Comment (pchickey) [reply lucet-wiggle/generate/src/lib.rs:31]: yes, exactly. Its a bummer that the proc macro crates have to be structured like this :/

File: lucet-wasi/tests/test_helpers/mod.rs [hunk 4]
```diff
@@ -91,19 +93,20 @@ pub fn run<P: AsRef<Path>>(path: P, ctx: WasiCtx) -> Result<__wasi_exitcode_t, E
...
             .expect("termination yields an exitcode")),
         Err(e) => bail!("runtime error: {}", e),
     }
 }
 
 pub fn run_with_stdout<P: AsRef<Path>>(
     path: P,
-    ctx: WasiCtxBuilder,
-) -> Result<(__wasi_exitcode_t, String), Error> {
+    ctx: &mut WasiCtxBuilder,
```
[the remark this answers is not in the document]
Comment (pchickey) [reply lucet-wasi/tests/test_helpers/mod.rs:104]: `WasiCtxBuilder`'s methods were all changed upstream from `(self, arg) -> Self` style to `(&mut self, arg)` style. I believe this was done for the ease of exposing via ffi.

File: lucetc/Cargo.toml [hunk 5]
```diff
@@ -27,6 +27,8 @@ target-lexicon = "0.10"
 lucet-module = { path = "../lucet-module", version = "=0.6.2-dev" }
 lucet-validate = { path = "../lucet-validate", version = "=0.6.2-dev" }
 lucet-wiggle-generate = { path = "../lucet-wiggle/generate", version = "=0.6.2-dev" }
+# lucet-wasi dep must have default-features off to avoid dep on lucet-runtime:
+lucet-wasi = { path = "../lucet-wasi", version = "=0.6.2-dev", default-features = false, features = [] }
```
[the remark this answers is not in the document]
Comment (pchickey) [reply lucetc/Cargo.toml:31]: Good point. That part of this PR was abandoned and I forgot to unwind the cargo features i put in place for it.

File: lucet-wasi/tests/test_helpers/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (acfoltzer) [reply lucet-wasi/tests/test_helpers/mod.rs:104]: Ah, we should still be able to pass the owned type here as long as it's declared `mut ctx: WasiCtxBuilder`. But again, not worth changing all the code that passes a `&mut ctx` now

File: lucet-wasi/src/lib.rs [hunk 6]
```diff
@@ -1,15 +1,61 @@
...
+pub use wasi_common::{WasiCtx, WasiCtxBuilder, WasiCtxBuilderError};
 
 // Wasi executables export the following symbol for the entry point:
 pub const START_SYMBOL: &str = "_start";
+
+pub fn bindings() -> lucet_module::bindings::Bindings {
+    lucet_wiggle_generate::bindings(&wasi_common::wasi::metadata::document())
+}
+
+pub fn document() -> wiggle::witx::Document {
```
[the remark this answers is not in the document]
Comment (pchickey) [reply lucet-wasi/src/lib.rs]: done. Using these methods in the tests has made them less fragile to changes in the filesystem layout.

File: lucet-wasi/src/runtime.rs [hunk 7]
```diff
@@ -0,0 +1,377 @@
+use lucet_runtime::{lucet_hostcall_terminate, vmctx::Vmctx};
+use std::cell::Ref;
+use wasi_common::wasi::wasi_snapshot_preview1::WasiSnapshotPreview1;
+use wasi_common::WasiCtx;
+use wiggle::{GuestError, GuestPtr};
+
+lucet_wasi_generate::bindings!({
+    ctx: LucetWasiCtx,
+    constructor: { LucetWasiCtx { vmctx } }
```
[the remark this answers is not in the document]
Comment (acfoltzer) [reply lucet-wasi/src/runtime.rs:9]: I saw you re-requested review; I think this is the last remaining request I had

**Pull request merged**